enhancements/storage/storage-hypershift.md

```diff
@@ -0,0 +1,1050 @@
+    to be able to run the controller parts on any cloud or on bare metal.
+  * vSphere is often behind a firewall and is often not accessible from outside of vSphere itself. I.e. if the
+    management cluster is outside of vSphere, it may be necessary to open new firewall ports.  
+* On vSphere, cluster-storage-operator installs vsphere-problem-detector. It's not clear if it should run in the
+  management cluster or in the guest cluster or in both. Because both cluster will need to talk to vSphere API
+  and the problem detector is there to check for common problems in the connection.
+
+### Risks and Mitigations
+
+#### Storage CRs are in the guest cluster
+
+* Since CRs for cluster-storage-operator, aws-ebs-csi-driver-operator and cluster-csi-snapshot-controller-operator are
+  in the guest cluster, a Cluster Instance Admin can DOS the whole CSI driver / CSI snapshots by editing these CRs in
+  an endless loop. The operators will redeploy the operands with each CR change and the new operands will need to wait
+  for leader election to expire, while the old operands are already terminated.
+  * We mitigate this issue a bit by overwriting user changes in the operator CRs by Hosted Cluster Config Operator that
+    runs in HyperShift management cluster. Still, an evil Cluster Instance Admin can easily write a loop that updates
+   the CRs in a busy loop.
+
+* Since leader election leased of the operands (AWS EBS CSI driver sidecars and csi-snapshot-controller) are in the
+  guest cluster, an evil Cluster Instance Admin can DOS the whole CSI driver / CSI snapshots by holding these leases
+  forever.
+  * To mitigate this we would need to teach upstream CSI sidecars to handle two kubeconfigs. One for leader election
+    in the management cluster and second to handle the actual storage objects in the guest cluster. There is no
+    such plan today upstream.
+
+* Due to our operator design based on library-go controllers, the operators use many conditions and overall
+  ClusterOperator conditions are computed as their union. A Cluster Instance Admin may add a condition
+  to `storage.status`, `clustercsidriver.status` or `csisnapshotcontroller.status` that would be translated by the
+  operators to overall `clusterOperator.status` of the corresponding operator.
+  * For example, adding condition `ExplitDegraded: true` to `storage.status` will render ClusterOperator `storage`
+    `Degraded: true`. cluster-storage-operator will not sanitize it.
+  * Fixing this is hard in library-go, as we use really a lot of conditions and creating a whitelist of the "official
+    ones" would be error-prone.
+
+All three issues are not specific to HyperShift, they exist in standalone OCP too. Fixing all these issues is IMO
+impossible if there is an evil Cluster Instance Admin. If we still want to fix it properly in a future
+OCP release, we may need to redesign storage operators completely! See Alternatives below.
```

Comment:
@derekwaynecarr, how probable is it that you will want us to move the CRs and leader election out of the guest cluster? If we will need to redesign everything in a future OCP release, then I'd prefer to start with the right design right away, even if it means more work now.

Reply:
Discussed outside of github. It seems to be an acceptable risk - guest cluster admins can break the guest cluster in many ways already.